In [2]:
from string import punctuation
from re import sub
import pandas as pd

## Дистрибуция во всему IAVD

In [3]:
iavd = pd.read_csv('all_iavd.csv')

In [4]:
# переведем в нижний регистр
iavd['entry'] = iavd['entry'].str.lower()

# удалим случаи, когда отдельно основа и корень
iavd = iavd[~iavd['entry'].str.endswith('-')]

### Слова с гортанной смычкой

In [5]:
iavd_words = iavd['entry']
iavd_words_with_gs = iavd_words[iavd_words.str.contains('ʔ')]

Всего входов с гортанной смычкой:

In [6]:
len(iavd_words_with_gs)

844

Два раза гортанная смычка:

In [7]:
# два раза гортанная смычка
iavd_words_with_gs.str.contains('ʔ.+ʔ').sum()

np.int64(30)

Три раза:

In [7]:
iavd_words_with_gs[iavd_words_with_gs.str.contains('ʔ.+ʔ.+ʔ')]

Series([], Name: entry, dtype: str)

Всего 844+30=874 гортанных смычки

#### Обработка двойных

Каждый вход с двумя гортанными смычками раздваиваем и одно из ее появлений меняем на GS чтобы правильно считать дистрибуцию

In [8]:
iavd_double_gs = iavd_words_with_gs[iavd_words_with_gs.str.contains('ʔ.+ʔ')]
iavd_words_with_gs_corrected = iavd_words_with_gs[~iavd_words_with_gs.str.contains('ʔ.+ʔ')]
iavd_words_with_gs_corrected = pd.concat((iavd_words_with_gs_corrected, iavd_double_gs.apply(lambda x: sub(r'(ʔ.+)ʔ', r'\1GS', x))))
iavd_words_with_gs_corrected = pd.concat((iavd_words_with_gs_corrected, iavd_double_gs.apply(lambda x: sub(r'ʔ(.+ʔ)', r'GS\1', x))))

In [9]:
iavd_words_with_gs_corrected

14         авазэʔԓкас
38         автобусаʔн
45         авэлаʔԓкас
55          азлаʔԓкас
69           айванʔан
            ...      
5647    ӽэGSлучэʔԓкас
5903     чэGSнӈэʔԓкас
6975     элGSэлэʔԓкас
7263      эGSрумʔэрум
7271       ˚аGSасхаʔн
Name: entry, Length: 874, dtype: str

Всего вхождений гортанной смычки:

In [10]:
LEN = len(iavd_words_with_gs_corrected)
LEN

874

### Что после гортанной смычки

Сонанты:

In [ ]:
iavd_words_with_gs_corrected.str.contains('ʔ[нймњӈлрљв]').sum() / LEN

np.float64(0.32036613272311215)

Гласные:

In [ ]:
iavd_words_with_gs_corrected.str.contains('ʔ[аиуоәэые]').sum() / LEN

np.float64(0.23684210526315788)

ԓ:

In [13]:
iavd_words_with_gs_corrected.str.contains('ʔԓ').sum() / LEN

np.float64(0.32379862700228834)

Посмотрим что встречается с ԓ:

In [14]:
pd.options.display.max_rows = 300
iavd_words_with_gs_corrected[iavd_words_with_gs_corrected.str.contains('ʔԓ')]

14            авазэʔԓкас
45            авэлаʔԓкас
55             азлаʔԓкас
79           айтатэʔԓкас
95            аӄтӽэʔԓкас
114        алупткʼаʔԓкас
116        алускʼваʔԓкас
120           алхпаʔԓкас
124      аләскʼваʔԓаткас
126        аләскʼваʔԓкас
153      ампәлсхэнаʔԓкас
161         ампэлаʔԓкаԓӽ
162          ампэлаʔԓкас
183         анањчпаʔԓкас
196            анзаʔԓкас
198          анзаʔԓсткас
205       анӄэӄчʼоʔԓкаԓӽ
218          ансаваʔԓкас
220          ансоӈаʔԓкас
237            ансэʔԓкас
238           ансэʔԓкиԓӽ
239            ансэʔԓкэс
240            ансэʔԓном
242          ансэʔԓэткас
247           анфчоʔԓкас
250          анхтзоʔԓкас
254          анхчоʔԓкаԓӽ
257         анчанзоʔԓкас
261        анәњчхчаʔԓкас
262       анәњчхчаʔԓкиԓӽ
263        анәњчхчаʔԓном
266        анэкстʼаʔԓкас
267         анэчпаʔԓкиԓӽ
272           ањԓнэʔԓкас
273          ањԓнэʔԓкиԓӽ
279            ањԓоʔԓкас
280           ањԓоʔԓкнэн
284     ањԓэнмайаӈаʔԓкас
316            аӈйаʔԓкас
384        атԓэмкʼаʔԓкас


На конце слова:

In [11]:
iavd_words_with_gs_corrected.str.endswith('ʔ').sum() / LEN

np.float64(0.09610983981693363)

Остальные:

In [303]:
iavd_words_with_gs_corrected[~(iavd_words_with_gs_corrected.str.endswith('ʔ') | iavd_words_with_gs_corrected.str.contains('ʔԓ') | iavd_words_with_gs_corrected.str.contains('ʔ[аиуоәэые]') | iavd_words_with_gs_corrected.str.contains('ʔ[нймњӈлрљв]'))]

486      ӑӈӄанэʔсх
1175       исхэʔсх
1400        каʔкас
1913        кэʔкʼф
2064       кʼэʔкас
2127          ӄаʔт
2128     ӄаʔт итʼэ
2129         ӄаʔтх
2130     ӄаʔтʼанан
2246      ӄәнʔтʼсх
2389     лаӽсхэʔсх
2392     лаӽсӽэʔсх
2558     лэӽсӽэʔсх
3331     нэԓʔчалас
3596       оʔтэкас
4664    толкʼаʔкас
4694        тоʔсэс
5644       ӽэʔӄʼԓх
6878     әӈӄанэʔсх
2131      ӄаʔтGSин
Name: entry, dtype: str

### Не встречается ни разу с

In [ ]:
# создаем набор символов
chars_vocab = set()
for entry in iavd['entry']:
    chars_vocab.update(set(entry))

chars_vocab = chars_vocab - set(punctuation) - set((' ', '˚', 'ь', 'ъ', 'ʼ', '̆')) # убираем явно нерелевантные символы

# правый контекст
print('Правый контекст: ', end='')
for char in chars_vocab:
    if iavd_words_with_gs_corrected.str.contains(f'ʔ{char}').sum() == 0:
        print(char, end=', ')

print()

# левый контекст
print('Левый контекст: ', end='')
for char in chars_vocab:
    if iavd_words_with_gs_corrected.str.contains(f'{char}ʔ').sum() == 0:
        print(char, end=', ')

Правый контекст: ё, д, г, ж, з, ф, щ, п, ы, ӽ, ӑ, ŏ, е, ц, c, ш, ў, ʔ, я, б, ю, х, 
Левый контекст: ё, д, г, ж, щ, п, ӑ, ŏ, е, ц, њ, c, ӄ, ш, ў, ʔ, я, б, ю, 

In [13]:
r = set(('д', 'х', 'п', 'б', 'з', 'ц', 'ʔ', 'ю', 'ӽ', 'ш', 'ў', 'щ', 'е', 'ф', 'г', 'ŏ', 'ж', 'ы', 'ё', 'я', 'ӑ'))
l = set(('д', 'ӄ', 'п', 'б', 'ц', 'ʔ', 'ю', 'ш', 'ў', 'щ', 'е', 'г', 'ŏ', 'њ', 'ж', 'ё', 'я', 'ӑ'))
l - r, r - l

({'њ', 'ӄ'}, {'з', 'ф', 'х', 'ы', 'ӽ'})

с лишнее -- оно латинское:

In [14]:
iavd[iavd['entry'].str.contains('c')]

,entry,link
269,анʔавнаc,/dictionary/ru/8747
844,зәмчʼсхэнәc,/dictionary/ru/3122


Абруптивы (они двумя символами):

In [80]:
iavd[iavd['entry'].str.contains('ʼʔ')]

,entry,link
1229,ичʼʔин,/dictionary/ru/13560


In [ ]:
iavd[iavd['entry'].str.contains('ʔ.ʼ')]

,entry,link
1913,кэʔкʼф,/dictionary/ru/9728
2130,ӄаʔтʼанан,/dictionary/ru/5325
2246,ӄәнʔтʼсх,/dictionary/ru/14830
5644,ӽэʔӄʼԓх,/dictionary/ru/12489


### Что перед гортанной смычкой

Сонанты:

In [ ]:
iavd_words_with_gs_corrected.str.contains('[нймњӈлрљв]ʔ').sum() / LEN

np.float64(0.06864988558352403)

Гласные:

In [ ]:
iavd_words_with_gs_corrected.str.contains('[аиуоәэые]ʔ').sum() / LEN

np.float64(0.8546910755148741)

Начало:

In [16]:
iavd_words_with_gs_corrected[iavd_words_with_gs_corrected.str.startswith('ʔ')] # но по словарю это морфема

7265    ʔин
Name: entry, dtype: str

Остальное:

In [ ]:
iavd_words_with_gs_corrected[iavd_words_with_gs_corrected.str.contains('[^аиуоәэыенймњӈлрљв]ʔ')]

359             асʔамԓкас
360                 асʔас
413               атхʔатх
414            атхʔатхчан
1214              ифԓӽʔин
1219              иӽԓӽʔин
1229               ичʼʔин
1411           кӑннэнхʔал
1476             ккаӈԓʔан
1481              клачʔин
1484              клыԓʔин
1512             кԓмаԓʔан
1519            кԓфањчʔан
1530            кԓфәњчʔан
1768              куԓӽʔин
1794           кӽаӄањԓʔин
1796             кчинзʔин
1804               кчфʔин
1852          кәнннэнхʔал
1950            кʼилфсʔин
2066             кʔэнэзэс
2149               ӄиӽʔин
2234     ӄуфсизʔин кивэчӽ
2376              ламԓʔан
2485        лыԓи ксуњԓʔин
2587             љэњиԓʔан
2632              ԓваԓʔан
2949              манхʔал
2981             мимсхʔин
2992              мныԓʔан
3063            мэчʼахʔал
3234             ныныԓʔан
3331            нэԓʔчалас
3368        ӈимсхʔин пʼэч
3725               пԓӽʔин
3998              салхʔал
4102         сичʼэӈанхʔал
4419              сэнхʔал
4423        

Два х:

In [ ]:
iavd_words_with_gs_corrected.str.contains('[хӽ]ʔ').sum() 

np.int64(31)

Два х и не суффикс исходно-продольного падежа:

In [74]:
iavd_words_with_gs_corrected[iavd_words_with_gs_corrected.str.contains('[хӽ]ʔ') & ~iavd_words_with_gs_corrected.str.endswith('хʔал')]

413           атхʔатх
414        атхʔатхчан
1214          ифԓӽʔин
1219          иӽԓӽʔин
1768          куԓӽʔин
2149           ӄиӽʔин
2981         мимсхʔин
3368    ӈимсхʔин пʼэч
3725           пԓӽʔин
4691          тосхʔан
5216          утхʔутх
5782           чԓхʔин
6139           әлхʔэс
7183        этӽʔэлкас
Name: entry, dtype: str

ԓ:

In [54]:
iavd_words_with_gs_corrected[iavd_words_with_gs_corrected.str.contains('ԓʔ')]

1476            ккаӈԓʔан
1484             клыԓʔин
1512            кԓмаԓʔан
1794          кӽаӄањԓʔин
2376             ламԓʔан
2485       лыԓи ксуњԓʔин
2587            љэњиԓʔан
2632             ԓваԓʔан
2992             мныԓʔан
3234            ныныԓʔан
3331           нэԓʔчалас
5071             тʼаԓʔан
5523           хэԓваԓʔан
1551         кԓʔасаӈаGSн
1414     кбарабанэGSԓʔан
1503         кљэчэGSԓʔин
1808          кчʼэGSԓʔин
1931        кʼансэGSԓʔан
1932    кʼанхтʼԓӈаGSԓʔан
1933       кʼанхчоGSԓʔан
Name: entry, dtype: str

Остальное без двух х и ԓ:

In [ ]:
iavd_words_with_gs_corrected[iavd_words_with_gs_corrected.str.contains('[^аиуоәэыенймњӈлрљвхӽԓ]ʔ')]

359             асʔамԓкас
360                 асʔас
1229               ичʼʔин
1481              клачʔин
1519            кԓфањчʔан
1530            кԓфәњчʔан
1796             кчинзʔин
1804               кчфʔин
1950            кʼилфсʔин
2066             кʔэнэзэс
2234     ӄуфсизʔин кивэчӽ
5392              хањчʔан
5935           чʼаӽлэзʔан
178     аназʔан ӈэззэGSњч
2131             ӄаGSтʔин
Name: entry, dtype: str

### Или

In [309]:
iavd_words_with_gs[iavd_words_with_gs.str.contains('[^аиуоәэые]ʔ[^аиуоәэые]')]

142      амамʔйоӄ
2246     ӄәнʔтʼсх
3331    нэԓʔчалас
Name: entry, dtype: str